portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java

```diff
@@ -0,0 +1,187 @@
+package com.google.sps.data;
+
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+
+/**
+ * Models an implementation of the TimeRangeGroup model using LinkedList. This is a slightly
+ * modified version of what hollyyuqizheng wrote for the ArrayList implementation.
+ */
+public class LinkedListTimeRangeGroup implements TimeRangeGroup {
+
+  // This list of all the time ranges will be sorted by start time ascending
+  // in add and delete methods. The time ranges stored in this list are
+  // pair-wise disjoint at any moment.
+  public List<TimeRange> allTimeRanges;
+
+  /**
+   * Adds all the input time ranges to the list of all time ranges. Also sorts the list of all
+   * ranges in the constructor.
+   */
+  public LinkedListTimeRangeGroup(Collection<TimeRange> timeRanges) {
+    allTimeRanges = new LinkedList<TimeRange>();
+    timeRanges.forEach(
+        (range) -> {
+          addTimeRange(range);
+        });
+  }
+
+  /**
+   * Adds a new time range to the list. If the time range to add overlaps with any existing time
+   * range, the overlapping time ranges will be merged.
+   */
+  public void addTimeRange(TimeRange timeRange) {
+    // If the original allTimeRanges list is empty,
+    // this is the first time we add anything to the list,
+    // so simply add the new time range to the list and return.
+    if (allTimeRanges.isEmpty()) {
+      allTimeRanges.add(timeRange);
+      return;
+    }
+    ListIterator<TimeRange> iterator = allTimeRanges.listIterator();
+
+    // This variable represents the latest time range previously exmamined.
+    // Initially, this variable points to the time range we want to add.
+    // As the for loop iterates through all the current time ranges already existing
+    // in allTimeRanges, this lastExaminedTimeRange variable is the time range with which
+    // any existing time range merges with, if necessary.
+    TimeRange lastExaminedTimeRange = timeRange;
+
+    while (iterator.hasNext()) {
+      TimeRange currentRange = iterator.next();
+
+      // If the current range is completely contained by the lastExaminedTimeRange,
+      // we need to remove the current range from the list as lastExaminedTimeRange
+      // will eventually replace it.
+      if (lastExaminedTimeRange.contains(currentRange)) {
+        iterator.remove();
+        continue;
```

Comment:
Consider the case where the initial internal time range list just has 1pm-2pm, and timeRange is 12pm-3pm. Then I think this line gets triggered, and the internal list stays at 1pm-2pm, rather than being updated to 12pm-3pm.

Reply:
Fixed. I had to rewrite a bit of the code but all of these corner cases having to due with a single element or adding one element all the way at the end work fine now. Tests were also added for these cases.